// @testing-library/angular/jest-utils/esm2020/src/lib/create-mock.mjs
export function createMock(type) {
    const mock = {};
    function mockFunctions(proto) {
        if (!proto) {
            return;
        }
        for (const prop of Object.getOwnPropertyNames(proto)) {
            if (prop === 'constructor') {
                continue;
            }
            const descriptor = Object.getOwnPropertyDescriptor(proto, prop);
            if (typeof descriptor?.value === 'function') {
                mock[prop] = jest.fn();
            }
        }
        mockFunctions(Object.getPrototypeOf(proto));
    }
    mockFunctions(type.prototype);
    return mock;
}
export function createMockWithValues(type, values) {
    const mock = createMock(type);
    Object.entries(values).forEach(([field, value]) => {
        mock[field] = value;
    });
    return mock;
}
export function provideMock(type) {
    return {
        provide: type,
        useValue: createMock(type),
    };
}
export function provideMockWithValues(type, values) {
    return {
        provide: type,
        useValue: createMockWithValues(type, values),
    };
}
